### PR TITLE
TSM Queries

### PR DESCRIPTION
**Pull request by jwilder**
This adds query support back to the dev tsm engine.  I still need to add the deduping logic in the `FileStore`, but that will be addressed in subsequent PR.

### CODE REVIEW

**Comment by jwilder** [comment]
@otoolep @pauldix

File: tsdb/engine/tsm1/data_file.go [hunk 1]
```diff
@@ -1049,6 +1109,8 @@ func (f *fileAccessor) readAll(key string) ([]Value, error) {
 			pos = block.Offset
 		}
 
+		println(pos, block.Size)
```
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: Debug?

File: tsdb/engine/tsm1/engine.go [hunk 2]
```diff
@@ -332,11 +375,22 @@ func (t *devTx) WriteTo(w io.Writer) (n int64, err error) { panic("not implement
...
+		Next(key string, time time.Time) ([]Value, error)
+		Prev(key string, time time.Time) ([]Value, error)
+	}
+
+	series string
+	fields []string
+
 	cache         Values
-	position      int
+	cachePos      int
```
Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: :-) Yeah, better name.

File: tsdb/engine/tsm1/engine.go [hunk 3]
```diff
@@ -418,14 +491,35 @@ func (c *devCursor) read() (int64, interface{}) {
...
 	}
-	return c.cache[c.position].UnixNano(), c.cache[c.position].Value()
+	return c.cache[c.cachePos].UnixNano(), c.cache[c.cachePos].Value()
 }
 
 // nextTSM returns the next value from the TSM files.
 func (c *devCursor) nextTSM() (int64, interface{}) {
-	return tsdb.EOF, nil
+	if c.ascending {
+
```
Comment (otoolep) [inline tsdb/engine/tsm1/engine.go]: Nit: you want this blank line?

**Comment by pauldix** [comment]
For the cursor, why does `Next` take a time? Shouldn't the `SeekTo` be the only thing that takes time?

**Comment by jwilder** [comment]
@pauldix `Next` is not on the cursor.  It's a function on the `TSMReader` to find the next block after the given time.  The time that cursor is at is held within the cursor.

**Comment by pauldix** [comment]
@jwilder ah ok, was confusing the two.

**Comment by otoolep** [comment]
Yeah, was confused by this too. Cool.

**Comment by pauldix** [comment]
Maybe it should be renamed to `NextBlock`?

File: tsdb/engine/tsm1/file_store.go [hunk 4]
```diff
@@ -261,6 +267,53 @@ func (f *FileStore) Read(key string, t time.Time) ([]Value, error) {
...
+	for _, fd := range f.files {
+		// Can this file possibly contain this key and timestamp?
+		if !fd.Contains(key) {
+			continue
+		}
+
+		// May have the key and time we are looking for so try to find
+		v, err := fd.Next(key, t)
+		if err != nil {
+			return nil, err
```
Comment (otoolep) [inline tsdb/engine/tsm1/file_store.go]: Just so I am clear, this is where dedupe logic would need to go?

File: tsdb/engine/tsm1/data_file.go [hunk 5]
```diff
@@ -855,6 +856,60 @@ func (t *TSMReader) Key(index int) string {
...
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	entries := t.index.Entries(key)
+	var entry *IndexEntry
+
+	if timestamp.Before(entries[0].MinTime) {
+		entry = entries[0]
+	} else {
+		for i, e := range entries {
```
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: In english, what is going on here?

File: tsdb/engine/tsm1/file_store.go (hunk 4 above)
Comment (jwilder) [inline tsdb/engine/tsm1/file_store.go]: Yes. This func `Prev` and `Read`.

File: tsdb/engine/tsm1/data_file.go (hunk 5 above)
Comment (jwilder) [inline tsdb/engine/tsm1/data_file.go]: This is look for the next block based on where timestamp resides.  L866 checks to see if the timestamps is before any block in the file.  If it is, we use the first block in the file.  Other wise, we find the block where the timestamp resides and use the next one after it.

**Comment by otoolep** [comment]
Yeah, that might be helpful. Pretty strong convention with regards to
`Next()` in our existing code.

**Comment by jwilder** [comment]
@pauldix @otoolep How about `ScanForward` and `ScanReverse`?  Or possibly `Scan(ascending bool)` and don't have `Next` and `Prev` be public.

`NextBlock` would mean I need to change `Read` and `ReadAll` to `ReadBlock` and `ReadAllBlocks` to make the naming consistent.

File: tsdb/engine/tsm1/data_file.go (hunk 5 above)
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: Ah yes, clear when you explain it -- thanks.

**Comment by otoolep** [comment]
Change makes sense to me, I'm OK with this going in.

**Comment by otoolep** [comment]
@jwilder `Scan` works for me too.

**Comment by pauldix** [comment]
@jwilder sure, that makes sense

**Comment by jwilder** [comment]
All comments addressed.

**Comment by pauldix** [comment]
:ship:

**Pull request merged**